### PR TITLE
[core] Fix autoscaler v2 for gcs FT

### PR DESCRIPTION
**Pull request by vitsai**
When recovering from Redis, we don't repopulate state in GcsAutoscalerStateManager (as it was stateless before). Since it is now stateful, fix this.

Verified that the added test fails on master.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rickyyx — APPROVED** [review]
Thanks for the quick fix! 

Does test_autoscaler_init fail if we don't do the change?

**Review by jjyao — APPROVED** [review]

**Comment by vitsai** [comment]
Yes, it does fail!

File: src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc [hunk 1]
```diff
@@ -231,6 +231,14 @@ GcsAutoscalerStateManager::GetAggregatedResourceLoad() const {
   return aggregate_load;
 };
 
+void GcsAutoscalerStateManager::Initialize(const GcsInitData &gcs_init_data) {
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:234]: Looks like GcsPlacementGroupManager also initializes from the gcs_init_data, so I guess there could be inconsistency before the first report from syncer to populate the actual node resources: 

- GcsAutoscaeler would proabbly think all nodes are empty after this `Initalize`
- But GcsPlacementGroupManager thinks some placement groups are already scheduled on the worker nodes (which they are) 

Unless we persist the cached autoscaler as part of init_data, but I feel this is probably fine.

File: python/ray/tests/test_gcs_fault_tolerance.py [hunk 2]
```diff
@@ -104,6 +104,56 @@ def test_gcs_server_restart_during_actor_creation(
...
+        ),
+    ],
+    indirect=True,
+)
+def test_autoscaler_init(
+    ray_start_cluster_head_with_external_redis,
+):
+    """Checks that the node failure detector is correct when gcs server restart.
+
+    We set the cluster to timeout nodes after 2 seconds of heartbeats. We then
```
Comment (rickyyx) [inline python/ray/tests/test_gcs_fault_tolerance.py:126]: nit: this comment looks unrelated?

File: python/ray/tests/test_gcs_fault_tolerance.py [hunk 3]
```diff
@@ -104,6 +104,56 @@ def test_gcs_server_restart_during_actor_creation(
...
+    assert nodes[0]["alive"] and nodes[1]["alive"]
+
+    head_node = cluster.head_node
+    gcs_server_process = head_node.all_processes["gcs_server"][0].process
+    gcs_server_pid = gcs_server_process.pid
+    # Kill gcs server.
+    cluster.head_node.kill_gcs_server()
+    # Wait to prevent the gcs server process becoming zombie.
+    gcs_server_process.wait()
+    wait_for_pid_to_exit(gcs_server_pid, 1000)
```
Comment (aslonnie) [inline python/ray/tests/test_gcs_fault_tolerance.py:146]: this it 1 second? this might to too tight, and creates flaky tests on CI.

maybe change it to 10sec or something.

**Comment by rickyyx** [comment]
nit: lint as well :)

File: python/ray/tests/test_gcs_fault_tolerance.py (hunk 3 above)
Comment (aslonnie) [reply python/ray/tests/test_gcs_fault_tolerance.py:146]: nevermind, this seems to be 1000 sec..

Comment (vitsai) [reply python/ray/tests/test_gcs_fault_tolerance.py]: changed it to 300

File: src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc (hunk 1 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:234]: GcsAutoscalerStateManager does not maintain placement group state, though? It uses `GcsPlacementGroupManager` as the source of truth

Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:234]: So any reported state will combine the two

**Comment by vitsai** [comment]
Pre-existing broken tests:
- test_memory_pressure
- ray_oom_prevention
- test_reference_counting

File: src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc (hunk 1 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:234]: yes, but GcsAutoscalerStateManager has the available resources info. But I think with PGmanager trying to rescheudle PGs, it should be synced before GCS reports to Autoscaler.
